mirrord-layer/src/tcp_mirror.rs

```diff
@@ -0,0 +1,170 @@
+use tokio::{
+    io::{AsyncReadExt, AsyncWriteExt},
+    net::TcpStream,
+    select,
+    sync::mpsc::{channel, Receiver, Sender},
+    task,
+};
+use tokio_stream::{wrappers::ReceiverStream, StreamExt};
+use tracing::{debug, error};
+
+use crate::{common::Listen, tcp::TCPHandler};
+
+async fn tcp_tunnel(mut local_stream: TcpStream, remote_stream: Receiver<Vec<u8>>) {
+    let mut remote_stream = ReceiverStream::new(remote_stream);
+    let mut buffer = vec![0; 1024];
+    loop {
+        select! {
+            message = remote_stream.next() => {
+                match message {
+                    Some(data) => {
+                        match local_stream.write_all(&data).await {
+                            Ok(_) => {},
+                            Err(err) => {error!("writing to local stream err {err:?}"); break;}
+                        }
+                    },
+                    None => {
+                        debug!("tcp tunnel exiting due to remote stream closed");
+                        break;
+                    }
+                }
+            },
+            res = local_stream.read(&mut buffer) => {
+                match res {
+                    Err(ref err) if err.kind() == std::io::ErrorKind::WouldBlock => {
+                        continue;
+                        },
+                    Err(err) => {
+                        error!("local stream received an error reading {err:?}");
+                        break;
+                    }
+                    Ok(n) if n == 0 => {
+                        debug!("tcp tunnel exiting due to local stream closed");
+                        break;
+                    },
+                    Ok(_) => {}
+                }
+
+            }
+        }
+    }
+    debug!("exiting tcp tunnel");
+}
+
+struct Connection {
+    writer: Sender<Vec<u8>>,
+    id: ConnectionID,
+}
+
+impl Eq for Connection {}
+
+impl PartialEq for Connection {
+    fn eq(&self, other: &Connection) -> bool {
+        self.id == other.id
+    }
+}
+
+impl Hash for Connection {
+    fn hash<H: Hasher>(&self, state: &mut H) {
+        self.id.hash(state);
+    }
+}
+
+impl Connection {
+    pub fn new(id: ConnectionID, writer: Sender<Vec<u8>>) -> Self {
+        Self { id, writer }
+    }
+
+    pub async fn write(&mut self, data: Vec<u8>) -> Result<()> {
+        Ok(self.writer.send(data).await?)
+    }
+}
+
+impl Borrow<ConnectionID> for Connection {
+    fn borrow(&self) -> &ConnectionID {
+        &self.id
+    }
+}
+
+/// Handles traffic mirroring
+pub struct TCPMirrorHandler {
+    ports: HashSet<Listen>,
+    running: bool,
+    connections: HashSet<Connection>,
+}
+
+#[async_trait]
+impl TCPHandler for TCPMirrorHandler {
+    fn new() -> TCPMirrorHandler {
+        Self {
+            ports: HashSet::new(),
+            running: true,
+            connections: HashSet::new(),
+        }
+    }
+
+    fn is_running(&mut self) -> bool {
```

Comment:
Why aren't the x_running functions implemented in the parent class?

Reply:
Actually most of the other functions too. Shouldn't the only mirror-specific code be in tcp_tunnel?